File: go/program/ops/opArrayConstructor.go
package ops

import (
	"github.com/2manyvcos/jpl/go/definition"
	"github.com/2manyvcos/jpl/go/jpl"
)

type opArrayConstructor struct{}

// { pipe: [op] }
func (opArrayConstructor) OP(runtime jpl.JPLRuntime, input any, params definition.JPLInstructionParams, scope jpl.JPLRuntimeScope, next jpl.JPLScopedPiper) ([]any, jpl.JPLError) {
	outputs, err := runtime.ExecuteInstructions(params.Pipe, []any{input}, scope, nil)
	if err != nil {
		return nil, err
	}
	return next.Pipe(outputs, scope)
}

// { pipe: function }
func (opArrayConstructor) Map(runtime jpl.JPLRuntime, params jpl.JPLInstructionParams) (definition.JPLInstructionParams, jpl.JPLError) {
	return definition.JPLInstructionParams{
		Pipe: call(params.Pipe),
	}, nil
}

File: js/lib/program/ops/opArrayConstructor.ts
import type { JPLRuntimeScope } from '@/library';
import type { JPLInstructionParams } from '@/library/definition';
import type JPLRuntime from '@/runtime';
import type { UserJPLInstructionParams } from '../params';
import { call } from './utils';

export default {
  /** { pipe: [op] } */
  async op(
    runtime: JPLRuntime,
    input: unknown,
    params: JPLInstructionParams,
    scope: JPLRuntimeScope,
    next: (
      output: unknown,
      scope: JPLRuntimeScope,
    ) => Promise<unknown[]> | unknown[],
  ): Promise<unknown[]> {
    const outputs = await runtime.executeInstructions(
      params.pipe ?? [],
      [input],
      scope,
    );
    return next(outputs, scope);
  },

  /** { pipe: function } */
  map(
    _runtime: JPLRuntime,
    params: UserJPLInstructionParams,
  ): JPLInstructionParams {
    return {
      pipe: call(params.pipe),
    };
  },
};
